File: ts/src/batch.ts
/**
 * Batch: accumulates writes and removes, commits once on commit().
 */

import git from 'isomorphic-git';
import {
  MODE_BLOB,
  MODE_LINK,
  MODE_TREE,
  FileNotFoundError,
  IsADirectoryError,
  PermissionError,
  BatchClosedError,
  type FsModule,
} from './types.js';
import { normalizePath } from './paths.js';
import { modeFromDisk, walkTo, existsAtPath, type TreeWrite } from './tree.js';

import { BatchWriter } from './fileobj.js';
import type { FS } from './fs.js';
import type { GitStore } from './gitstore.js';

/**
 * Accumulates writes and removes, then commits all changes atomically
 * when `commit()` is called.
 *
 * Nothing is committed if the batch is not explicitly committed.
 * Use `commit()` to finalize all staged changes in a single commit.
 */
export class Batch {
  private _fs: FS;
  private _store: GitStore;
  private _fsModule: FsModule;
  private _gitdir: string;
  private _message: string | null;
  private _operation: string | null;
  private _parents: FS[] | undefined;
  private _writes = new Map<string, TreeWrite>();
  private _removes = new Set<string>();
  private _closed = false;

  /** The resulting FS snapshot after commit. Null until commit() completes. */
  fs: FS | null = null;

  constructor(fs: FS, message?: string | null, operation?: string | null, parents?: FS[]) {
    if (!fs._writable) {
      throw new PermissionError('Cannot batch on a read-only snapshot');
    }
    this._fs = fs;
    this._store = fs._store;
    this._fsModule = fs._store._fsModule;
    this._gitdir = fs._store._gitdir;
    this._message = message ?? null;
    this._operation = operation ?? null;
    this._parents = parents;
  }

  private _checkOpen(): void {
    if (this._closed) throw new BatchClosedError('Batch is closed');
  }

  /**
   * Stage a file write. Creates the blob immediately in the object store.
   *
   * @param path - Destination path in the repo.
   * @param data - Raw bytes to write.
   * @param opts.mode - File mode override (e.g. MODE_BLOB_EXEC for executable).
   */
  async write(path: string, data: Uint8Array, opts?: { mode?: string }): Promise<void> {
    this._checkOpen();
    const normalized = normalizePath(path);
    this._removes.delete(normalized);
    const blobOid = await git.writeBlob({ fs: this._fsModule, gitdir: this._gitdir, blob: data });
    this._writes.set(normalized, {
      oid: blobOid,
      mode: opts?.mode ?? MODE_BLOB,
    });
  }

  /**
   * Stage a write from a local file. Reads the file and creates the blob.
   *
   * Executable permission is auto-detected from disk unless `opts.mode` is set.
   *
   * @param path - Destination path in the repo.
   * @param localPath - Path to the local file.
   * @param opts.mode - File mode override (e.g. MODE_BLOB_EXEC for executable).
   */
  async writeFromFile(
    path: string,
    localPath: string,
    opts?: { mode?: string },
  ): Promise<void> {
    this._checkOpen();
    const normalized = normalizePath(path);
    this._removes.delete(normalized);

    const detectedMode = await modeFromDisk(this._fsModule, localPath);
    const mode = opts?.mode ?? detectedMode;
    const data = (await this._fsModule.promises.readFile(localPath)) as Uint8Array;
    const blobOid = await git.writeBlob({ fs: this._fsModule, gitdir: this._gitdir, blob: data });
    this._writes.set(normalized, { oid: blobOid, mode });
  }

  /**
   * Stage a symbolic link entry.
   *
   * @param path - Symlink path in the repo.
   * @param target - The symlink target string.
   */
  async writeSymlink(path: string, target: string): Promise<void> {
    this._checkOpen();
    const normalized = normalizePath(path);
    this._removes.delete(normalized);
    const data = new TextEncoder().encode(target);
    const blobOid = await git.writeBlob({ fs: this._fsModule, gitdir: this._gitdir, blob: data });
    this._writes.set(normalized, { oid: blobOid, mode: MODE_LINK });
  }

  /**
   * Stage a file removal.
   *
   * @param path - Path to remove from the repo.
   * @throws {FileNotFoundError} If the path does not exist in the repo or pending writes.
   * @throws {IsADirectoryError} If the path is a directory.
   */
  async remove(path: string): Promise<void> {
    this._checkOpen();
    const normalized = normalizePath(path);
    const pendingWrite = this._writes.has(normalized);
    const existsInBase = await existsAtPath(
      this._fsModule,
      this._gitdir,
      this._fs._treeOid,
      normalized,
    );

    if (!pendingWrite && !existsInBase) {
      throw new FileNotFoundError(normalized);
    }

    // Don't allow removing directories
    if (existsInBase) {
      const entry = await walkTo(this._fsModule, this._gitdir, this._fs._treeOid, normalized);
      if (entry.mode === MODE_TREE) {
        throw new IsADirectoryError(normalized);
      }
    }

    this._writes.delete(normalized);
    if (existsInBase) {
      this._removes.add(normalized);
    }
  }

  /**
   * Return a buffered writer that stages to this batch on close.
   *
   * Accepts `Uint8Array` or `string` via `write()`. Strings are UTF-8 encoded.
   *
   * @param path - Destination path in the repo.
   * @returns A BatchWriter instance. Call `close()` to flush and stage.
   */
  writer(path: string): BatchWriter {
    this._checkOpen();
    return new BatchWriter(this, path);
  }

  /**
   * Commit all accumulated changes atomically.
   *
   * After calling this the batch is closed and no further writes are allowed.
   *
   * @returns The resulting FS snapshot.
   * @throws {StaleSnapshotError} If the branch has advanced since the snapshot.
   */
  async commit(): Promise<FS> {
    if (this._closed) throw new BatchClosedError('Batch is already committed');
    this._closed = true;

    if (this._writes.size === 0 && this._removes.size === 0) {
      this.fs = this._fs;
      return this._fs;
    }

    this.fs = await this._fs._commitChanges(
      this._writes,
      this._removes,
      this._message,
      this._operation,
      this._parents,
    );
    return this.fs;
  }
}

File: rs/src/batch.rs
use std::path::Path;

use crate::error::{Error, Result};
use crate::fs::{Fs, TreeWrite};
use crate::tree;
use crate::types::{MODE_BLOB, MODE_LINK};

/// Accumulates writes and removes, committing them atomically.
///
/// [`commit`](Batch::commit) takes ownership (`self`), so the compiler
/// enforces that no further writes happen after committing.
pub struct Batch {
    pub(crate) fs: Fs,
    pub(crate) writes: Vec<(String, Option<TreeWrite>)>,
    pub(crate) removes: Vec<String>,
    pub(crate) message: Option<String>,
    pub(crate) operation: Option<String>,
    pub(crate) parents: Vec<Fs>,
    pub(crate) closed: bool,
}

impl Batch {
    fn require_open(&self) -> Result<()> {
        if self.closed {
            Err(Error::BatchClosed)
        } else {
            Ok(())
        }
    }

    /// Write raw bytes to `path` with the default blob mode (`0o100644`).
    pub fn write(&mut self, path: &str, data: &[u8]) -> Result<()> {
        self.write_with_mode(path, data, MODE_BLOB)
    }

    /// Write raw bytes to `path` with an explicit file mode.
    pub fn write_with_mode(&mut self, path: &str, data: &[u8], mode: u32) -> Result<()> {
        self.require_open()?;
        let path = crate::paths::normalize_path(path)?;

        let tw = self.fs.with_repo(|repo| {
            let blob_oid = repo.blob(data).map_err(Error::git)?;
            Ok(TreeWrite {
                data: data.to_vec(),
                oid: blob_oid,
                mode,
            })
        })?;

        // Remove from removes if present
        self.removes.retain(|p| p != &path);
        // Remove existing write for same path
        self.writes.retain(|(p, _)| p != &path);
        self.writes.push((path, Some(tw)));
        Ok(())
    }

    /// Write the contents of a disk file to `path`.
    pub fn write_from_file(&mut self, path: &str, src: &Path) -> Result<()> {
        self.require_open()?;
        let data = std::fs::read(src).map_err(|e| Error::io(src, e))?;
        let mode = tree::mode_from_disk(src).unwrap_or(MODE_BLOB);
        self.write_with_mode(path, &data, mode)
    }

    /// Write a symlink at `path`.
    pub fn write_symlink(&mut self, path: &str, target: &str) -> Result<()> {
        self.require_open()?;
        let path = crate::paths::normalize_path(path)?;

        let tw = self.fs.with_repo(|repo| {
            let blob_oid = repo.blob(target.as_bytes()).map_err(Error::git)?;
            Ok(TreeWrite {
                data: target.as_bytes().to_vec(),
                oid: blob_oid,
                mode: MODE_LINK,
            })
        })?;

        self.removes.retain(|p| p != &path);
        self.writes.retain(|(p, _)| p != &path);
        self.writes.push((path, Some(tw)));
        Ok(())
    }

    /// Return a buffered [`BatchWriter`](crate::fileobj::BatchWriter) that
    /// stages to this batch on close.
    ///
    /// The writer implements [`std::io::Write`], so you can use `write_all` etc.
    pub fn writer(&mut self, path: &str) -> Result<crate::fileobj::BatchWriter<'_>> {
        self.require_open()?;
        let normalized = crate::paths::normalize_path(path)?;
        Ok(crate::fileobj::BatchWriter::new(self, normalized))
    }

    /// Mark `path` for removal.
    pub fn remove(&mut self, path: &str) -> Result<()> {
        self.require_open()?;
        let path = crate::paths::normalize_path(path)?;

        // Remove from writes if present
        self.writes.retain(|(p, _)| p != &path);

        if !self.removes.contains(&path) {
            self.removes.push(path);
        }
        Ok(())
    }

    /// Commit all accumulated writes and removes atomically.
    ///
    /// Consumes the `Batch` and returns the resulting [`Fs`] snapshot.
    /// If no writes or removes were recorded, the original `Fs` is
    /// returned unchanged (no empty commit is created).
    ///
    /// # Errors
    /// Returns an error if the batch is already closed or the commit fails.
    pub fn commit(mut self) -> Result<Fs> {
        self.closed = true;

        if self.writes.is_empty() && self.removes.is_empty() {
            return Ok(self.fs);
        }

        // Merge removes into writes as (path, None)
        let mut all_writes = self.writes;
        for path in self.removes {
            all_writes.push((path, None));
        }

        let message = self.message.unwrap_or_else(|| {
            crate::paths::format_commit_message(
                self.operation.as_deref().unwrap_or("batch"),
                None,
            )
        });

        let extra: Vec<&Fs> = self.parents.iter().collect();
        self.fs.commit_changes_with_parents(&all_writes, &message, &extra)
    }

    /// Returns `true` if `commit()` has already been called.
    pub fn is_closed(&self) -> bool {
        self.closed
    }
}
